demos/tutorials/JointDistribution.py

```diff
@@ -0,0 +1,233 @@
+"""
+Joint Distribution tutorial
+===========================
+
+    This tutorial shows how to use the Joint Distribution class.
+
+"""
+# %%
+# Setup
+# -----
+# We start by importing the necessary modules and loading the model and data.
+# The model is a simple 1D convolution, but any model can be used.
+# We also define some helper variables to make the code more readable.
+
+import sys; sys.path.append("..")
+import numpy as np
+from cuqi.distribution import GaussianCov, Gamma, JointDistribution
+from cuqi.testproblem import Deconvolution1D
+
+# Model and data
+A, y_obs, _ = Deconvolution1D.get_components()
+
+# Model dimensions
+n = A.domain_dim
+m = A.range_dim
+
+# "Idendity" matricies
+In = np.ones(n)
+Im = np.ones(m)
+
+# %%
+# Bayesian modelling with the Joint Distribution
+# ----------------------------------------------
+#
+# The joint distribution is a tool that allows us to define Bayesian models.
+# This is achieved by combining multiple distributions into a single object.
+#
+# For example, consider the following model
+#
+# .. math::
+#
+#   \mathbf{x} &\sim \mathcal{N}(\mathbf{0}, d^2 \mathbf{I}_n) \\
+#   \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, s^2 \mathbf{I}_m)
+#
+# where :math:`\mathbf{A}\in\mathbb{R}^{m\times n}` is a matrix that defines the linear convolution.
+# and :math:`d`, :math:`s` are fixed parameters for the prior and noise standard deviations respectively.
+#
+# We can write this model in CUQIpy as follows:
+
+# Define fixed parameters
+d = 0.1
+s = 0.05
+
+# Define distributions
+x = GaussianCov(np.zeros(n), d*In)
+y = GaussianCov(lambda x: A@x, s*Im)
+
+# Define joint distribution p(x,y)
+joint = JointDistribution([x, y])
+
+# The joint distributions prints a summary of its components.
+print(joint)
+
+# %%
+# Basic properties of the joint distribution
+# ------------------------------------------
+#
+# The joint distribution comes with two main properties:
+# dim (dimension) and geometry. Because its a joint distribution
+# over multiple variables, these are lists.
+
+print(joint.dim)
```

Comment:
The dimension of y is None, is that intended? Being defined through a lambda function its size is perhaps hard to know in general?

Reply:
I think we agreed you would create an issue on allowing the user to specify dimension manually?

Reply:
#28